File: nodejs/patterns/src/creational/abstract_factory.js
class LightButton {
  paint() {
    return "light-button";
  }
}

class DarkButton {
  paint() {
    return "dark-button";
  }
}

class LightCheckbox {
  paint() {
    return "light-checkbox";
  }
}

class DarkCheckbox {
  paint() {
    return "dark-checkbox";
  }
}

class LightThemeFactory {
  createButton() {
    return new LightButton();
  }

  createCheckbox() {
    return new LightCheckbox();
  }
}

class DarkThemeFactory {
  createButton() {
    return new DarkButton();
  }

  createCheckbox() {
    return new DarkCheckbox();
  }
}

function renderUi(factory) {
  return [factory.createButton().paint(), factory.createCheckbox().paint()];
}

function demo() {
  return renderUi(new DarkThemeFactory());
}

module.exports = { LightThemeFactory, DarkThemeFactory, demo };

File: python/patterns/src/prototype_patterns/creational/abstract_factory.py
"""Abstract Factory pattern example."""

from __future__ import annotations

from abc import ABC, abstractmethod


class Button(ABC):
    @abstractmethod
    def paint(self) -> str:
        raise NotImplementedError


class Checkbox(ABC):
    @abstractmethod
    def paint(self) -> str:
        raise NotImplementedError


class LightButton(Button):
    def paint(self) -> str:
        return "light-button"


class DarkButton(Button):
    def paint(self) -> str:
        return "dark-button"


class LightCheckbox(Checkbox):
    def paint(self) -> str:
        return "light-checkbox"


class DarkCheckbox(Checkbox):
    def paint(self) -> str:
        return "dark-checkbox"


class ThemeFactory(ABC):
    @abstractmethod
    def create_button(self) -> Button:
        raise NotImplementedError

    @abstractmethod
    def create_checkbox(self) -> Checkbox:
        raise NotImplementedError


class LightThemeFactory(ThemeFactory):
    def create_button(self) -> Button:
        return LightButton()

    def create_checkbox(self) -> Checkbox:
        return LightCheckbox()


class DarkThemeFactory(ThemeFactory):
    def create_button(self) -> Button:
        return DarkButton()

    def create_checkbox(self) -> Checkbox:
        return DarkCheckbox()


def render_ui(factory: ThemeFactory) -> tuple[str, str]:
    return (factory.create_button().paint(), factory.create_checkbox().paint())


def demo() -> tuple[str, str]:
    return render_ui(DarkThemeFactory())
